Multi-Asset Migration Algorithm

jonathanknowles opened the pull request:
# Issue Number

ADP-839

# Overview

This PR adds an algorithm for performing migrations of multi-asset UTxO sets from one wallet to another.

It consists of four main parts:

1. Module `Migration`, which provides a top-level public API, defined in terms of concrete wallet types.
2. Module `Migration.Planning`, which contains an algorithm for planning migrations at a high
level. It determines how to partition a UTxO set into entries of different types, and in which order to add UTxO entries to selections.
3. Module `Migration.Selection`, which provides functions for incrementally constructing a
selection to be included in a migration plan. (A selection is the basis for an individual transaction.)
4. The `TxConstraints` type, which provides an abstract cost model for transactions, allowing parts of a transaction can be costed individually. This allows the incremental cost of extending a selection to be calculated without having to recalculate the cost of the entire selection.

# Performance

For pure ada wallets, we can process around 30,000 UTxO entries per second. 

We can process around 3,000 UTxO entries per second with wallets of the following composition:

| Percentage | UTxO Entry Type |
| -- | -- |
| 10% | Pure ada entries below the minimum ada quantity |
| 40% | Pure ada entries with at least the minimum ada quantity |
| 40% | Multi-asset token bundles with precisely the minimum ada quantity |
| 10% | Multi-asset token bundles with more than the minimum ada quantity |

# Design

## UTxO partitioning

The algorithm starts by partitioning the UTxO set into three types of UTxO entry:

| Type | Definition |
| -- | -- |
| **_supporter_** | an entry that is capable of paying for itself |
| **_freerider_** | an entry that is **not** capable of paying for itself |
| **_ignorable_** | a dust ada entry that should not be included because its value is less than the marginal cost of an input. |

## Adding entries to transactions

The algorithm uses a **supporter** entry to initialize each transaction, but then gives priority to adding **freerider** entries until the transaction is full. It adds a **supporter** entry only when there's not enough ada available to pay for an additional **freerider** entry.

<img src="https://user-images.githubusercontent.com/206319/115514178-6c521b80-a2b6-11eb-808f-92b4c679b99a.png" alt="extending a selection" width="300px">

## Creating outputs within transactions

The algorithm minimizes the number of outputs for each transaction (and thus minimizes the fee required) by merging together all value provided by the inputs. Due to the nature of our transaction constraints (which limit the size of an output), the algorithm usually produces transactions with between 1 and 4 outputs.

# Requirements

The algorithm in this PR is designed to satisfy the following requirements:

-  **REQ-1**
  The algorithm _must_ succeed at selecting all pure ada entries greater than `txBaseFee` + `minAdaQuantity` + `marginalFeeForInputAndOutput`.
- **REQ-2**
  The algorithm _must_ succeed at selecting all multi-asset entries whose ada quantities are greater than or equal to `txBaseFee` + `minAdaQuantity` + `marginalFeeForInputAndOutput`.
- **REQ-3**
  The algorithm _should_ make a best effort to select entries whose ada quantities are less than `txBaseFee` + `minAdaQuantity` + `marginalFeeForInputAndOutput`, but is permitted to omit such entries if it cannot find a viable strategy to select them.
- **REQ-4**
  The algorithm _must_ produce **T**: a set of transactions for which fees are balanced, and **R**: the remainder of the UTxO set that could not be selected, such that the inputs of **T** and the entries in **R** are disjoint, and such that the union of inputs in **T** and entries in **R** is equal to the original UTxO set.
- **REQ-5**
  The algorithm _must_ preserve the balance of every non-ada asset, so that for any given asset **a**, the total quantity of asset **a** in the selected inputs is exactly equal to the total quantity of asset **a** in the generated outputs.
- **REQ-6**
  The algorithm _must not_ produce transactions with dependencies between them; it should be possible to broadcast the generated transactions in any order.
- **REQ-7**
  The algorithm _must not_ produce outputs with token quantities that exceed the maximum allowed in an output `(2^64 − 1)`.
- **REQ-8**
  The algorithm _must not_ produce outputs whose serialized length exceeds the maximum allowed length (`4000` bytes).
- **REQ-9**
  The algorithm _must not_ produce transactions that exceed the transaction size limit (`16384` bytes).
- **REQ-10**
  The algorithm _must_ move the reward balance in its entirety.
- **REQ-11**
  The algorithm _must_ be deterministic: if run twice on the same UTxO set, it must generate exactly the same migration plan. (This is necessary to support an API where a non-hardware wallet user can generate a plan, inspect the plan, and then initiate a migration without having to resubmit the plan.)

# Non-Requirements

- **NON-REQ-1**
  Preserving the UTxO distribution.

jonathanknowles committed:
Add `TxConstraints` type to `Primitive.Types.Tx`.

The `TxConstraints` type provides an abstract cost model for
transactions, allowing parts of a transaction can be costed
individually.

jonathanknowles committed:
Add module `Migration.Selection`.

This module contains functions for incrementally constructing a
selection to be included in a migration plan.

A selection is the basis for an individual transaction.

jonathanknowles committed:
Add module `Migration.Planning`.

This module contains an algorithm for planning migrations at a high
level. It determines how to partition the UTxO set into entries of
different types, and in which order to add UTxO entries to selections.

jonathanknowles committed:
Add module `Migration`.

This module provides a public API for planning migrations.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -0,0 +1,787 @@
+        --
+        -- As a tie-breaker, we give priority to outputs with smaller numbers
+        -- of assets. Merging with a smaller output is more likely to succeed,
+        -- because merging with a larger ouput is more likely to fall foul of
```
Small typo:
```suggestion
        -- because merging with a larger output is more likely to fall foul of
```

jonathanknowles replied:
Fixed in 9519bdd9d2aaa404821866dec171bd8557a2c05c.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Planning.hs:
```diff
@@ -0,0 +1,271 @@
+    -> UTxO
+    -> CategorizedUTxO (TxIn, TxOut)
+categorizeUTxO constraints (UTxO u) = categorizeUTxOEntries constraints $
+    (\(i, o) -> ((i, o), view #tokens o)) <$> Map.toList u
```
It seems `categorizeUTxO` and `uncategorizeUTxO` are both needed only in `createPlan`. 

How about moving the `entry <-> UTxO` conversion there as part of a `where`-clause? It might be a bit clearer division of responsibility then.

jonathanknowles replied:
> It seems categorizeUTxO and uncategorizeUTxO are both needed only in createPlan.
> How about moving the entry <-> UTxO conversion there as part of a where-clause? It might be a bit clearer division of responsibility then.


Good idea! I've fixed this in d4819813bff15429724003d79aa489a77327a1fa.

Anviking replied:
d4819813bff15429724003d79aa489a77327a1fa seems to have been dropped. Was this a mistake?

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -0,0 +1,787 @@
+    -> Selection i s
+    -> SelectionCorrectness s
+verify constraints selection
+    | Just e <- checkAssetBalance selection =
```
I hadn't seen this style before. Wouldn't 
```haskell
checkAssetBalance :: Selection i s -> Either SelectionAssetBalanceIncorrectError ()
```
and
```haskell
verify = do
    left SelectionAssetBalanceIncorrect $ checkAssetBalance selection
    left SelectionFeeIncorrect $ ...
    ...
```
be more plain?

jonathanknowles replied:
I agree that the control flow could be a little clearer.

What do you think about this solution: d31d1ec5a72202561fb4f15012ec6ba1dbf1d229?

Anviking replied:
I think it reads better, but I don't see why we need `SelectionCorrectness` https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r617352652

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/Migration/PlanningSpec.hs:
```diff
@@ -0,0 +1,367 @@
+        , mockRewardWithdrawal
+        }
+
+prop_createPlan :: ArgsForCreatePlan -> Property
```
I wonder: isn't this `ArgsX` pattern similar to `forAll` 🤔 

Haven't done this myself, but I think you could have 
```haskell
prop_createPlan :: MockTxConstraints -> MockWithdrawal -> Property
prop_createPlan constraints (MockWithdrawal wdrl) = forAll (genMockInputs constraints) $ \inputs -> 
```
such that the generators are only _as coupled_ as necessary but no more, and you can more easily have shrinkers for the constraints and the withdrawal.

jonathanknowles replied:
Hi @Anviking 

As requested, I've used `forAll` and `forAllShrink` to replace all the `ArgsX` types.

Thanks for the suggestion, I think it has really simplified the test code. (Much less boilerplate!)

Anviking replied:
Nice! 🎉

jonathanknowles committed:
Clarify the control flow of `Selection.verify`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r616027279

jonathanknowles committed:
Add support for pretty-printing arbitrary `Show` instances.

This change makes it possible to pretty-print any value that is an
instance of `Show` in property test failures.

jonathanknowles committed:
Make generator for `MockInputId` more collision-resistant.

We explicitly specify the full range of a `Word8` by using `choose`,
which guarantees to generate a uniform distribution.

Collisions were seen in test runs. After this change is applied,
collisions disappear.

jonathanknowles committed:
Add property test for `categorizeUTxOEntries`.

We demonstrate that `uncategorize . categorize == id`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#issuecomment-822413404

jonathanknowles committed:
Generate fewer examples in `prop_createPlan`.

This makes the overall test suite run time more reasonable.

jonathanknowles committed:
Use `tabulate` instead of `label` in `prop_createPlan`.

This produces a clearer statistical summary.

In addition, this commit replaces "percentage not selected" with
"percentage selected", which is arguably easier to interpret.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#issuecomment-822413404

jonathanknowles committed:
Add further input preservation test within `prop_createPlan`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#issuecomment-822413404

jonathanknowles committed:
Test that the public and private interfaces are equivalent.

This commit adds a simple sanity test to check that the public and
private interfaces for creating a migration plan are equivalent.

jonathanknowles committed:
Fix typos of "outputs" in comments.

This also fixes a similar typo in the ordinary coin selection algorithm.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r615821640

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -0,0 +1,796 @@
+-- Selections
+--------------------------------------------------------------------------------
+
+data Selection i s = Selection
```
Perhaps use more descriptive type parameters here.

Instead of `i`, use `inputId`.
Instead of `s`, use `size`.

Anviking replied:
I think `size` makes sense if keeping it 👍 

How about just `input` instead of `inputId`? I don't get why this distinction is preferable.

It seems `i ~ (TxIn, TxOut)` for the real algorithm also... Doesn't seem like an "id"?

jonathanknowles replied:
Okay, I've gone for `input` and `size`, as suggested.

Fixed in 2a2f17dc8384a68ff2e968f39a125d0a02f1e926.

> It seems i ~ (TxIn, TxOut) for the real algorithm also... Doesn't seem like an "id"?

Thinking about this, `createPlan` actually only needs to return a `MigrationPlan` where `i ~ TxIn`, since the caller of `createPlan` can figure out which `TxOut` each `TxIn` refers to, if they really need to do so.

Perhaps we can change this to be `i ~ TxIn`.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Migration/SelectionSpec.hs:
```diff
@@ -0,0 +1,1036 @@
+        "Failure due to oversized selection" $
+    case result of
+        Left SelectionAdaInsufficient ->
+            property True
```
Suggestion: instead of returning `True` here, call `Selection.create` to verify that this function will also produce the same error.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Migration/SelectionSpec.hs:
```diff
@@ -0,0 +1,1036 @@
+        Left SelectionAdaInsufficient ->
+            property True
+        Left (SelectionFull e) ->
+            property (selectionSizeMaximum e < selectionSizeRequired e)
```
Suggestion: also call `Selection.create` here to verify that this function will also produce the same error.

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/MigrationSpec.hs:
```diff
@@ -0,0 +1,174 @@
+    ( MockTxConstraints
+    , genCoinRange
+    , genMockTxConstraints
+    , genTokenBundleMixed
```
I think this is technically against our coding standards:

> [**Test files are separated and self-contained**](https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#test-files-are-separated-and-self-contained)
> It is really easy to make the testing code more complex than the actual code it's initially testing. Limiting the interaction between test modules helps keeping a good maintainability and a rather low overhead when it comes to extend, modify, read or comprehend some tests. Also, in many cases, we do actually want to have different arbitrary generators for different test cases so sharing instances is risky and cumbersome.

jonathanknowles replied:
> Limiting the interaction between test modules helps keeping a good maintainability and a rather low overhead when it comes to extend, modify, read or comprehend some tests.

I think there's a clear tension between **inter-module interaction** (which we usually want to minimize) and **code reuse** (which might be desirable).

I agree we **definitely** want to avoid the problem of "spooky action at a distance"! (Where we change something in one module that has a nefarious effect on another module.)

In the past, we had many `Arbitrary` instances that were shared between test modules. Of course, shared `Arbitrary` instances have the following disadvantages:
 - When looking at a given call to `arbitrary`, it's often hard to tell where the instance is defined. You first need to know the type, and then need to check the module dependency graph.
 - When changing the definition of an instance, it's hard to tell which other modules and tests will be affected.
 - Instances may be defined in terms of more calls to `arbitrary`, which makes solving the instance resolution graph even harder.
  
In this PR, we've not shared any `Arbitrary` instances between modules. (Except, perhaps, for the `Arbitrary (NonEmpty a)` instance, which is relatively innocuous.)

In every case of code sharing, we've explicitly imported just the generator functions we need. So it should be easy for a reader to figure out where something is coming from, and there should be no "spooky action at a distance".

In my mind, this is a good balance: we get the code reuse we want, and the reader can see explicitly where each piece of code comes from.

Having said this, I think we might end up (in the next PR), moving some of these generators to a `Gen` module, so that we can reuse them from the Shelley test suite.

jonathanknowles committed:
Use more descriptive type variables.

Use `size` to represent sizes.
Use `input` to represent inputs (identifiers).

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r616397493

jonathanknowles committed:
Add documentation comments to functions in `Planning` and `Selection`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -0,0 +1,801 @@
+    run (Coin 0) remaining processed =
+        (Coin 0, L.reverse processed <> remaining)
+    run feeExcessRemaining [] processed =
+        (feeExcessRemaining, L.reverse processed)
```
Suggestion: remove these reversals, as they are almost certainly unnecessary.

Their existence is a hangover from an earlier iteration of the design, where maintaining the order of outputs was important.

jonathanknowles replied:
Fixed in f9de1b4dad3c642b350217c0afad318ae058e9ec.

jonathanknowles committed:
Remove unnecessary reversals from `minimizeFee`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618/files#r617179637

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Migration/PlanningSpec.hs:
```diff
@@ -0,0 +1,418 @@
+
+instance Arbitrary (Giant ArgsForCreatePlan) where
+    arbitrary = Giant <$> genArgsForCreatePlan
+        (100_000, 100_000) genMockInputAdaOnly
```
With recent optimizations, I think we can relax the restriction of using ada-only inputs for the "giant" test, provided that we reduce the "giant" size to something more manageable, such as 10,000 entries. (Which should take around 3 seconds.)

jonathanknowles replied:
Fixed in 522a895a11384b9faebc6868997f08f836ef0fde.

jonathanknowles committed:
Relax the ada-only restriction when testing giant migrations.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618/files#r617241998

jonathanknowles committed:
Explicitly test the creation of empty migration plans.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -0,0 +1,956 @@
+--
+data SelectionCorrectness size
+    = SelectionCorrect
+    | SelectionIncorrect (SelectionCorrectnessError size)
```
Could we replace this with just `Either (SelectionCorrectnessError size) ()`? To me it doesn't seem worth it's own type.

Anviking replied:
Also, looking _now_, I wonder if having the correctness checks in this module actually makes sense. Maybe could be moved later, if an opportunity arises.

jonathanknowles replied:
> Could we replace this with just Either (SelectionCorrectnessError size) ()? To me it doesn't seem worth it's own type.

Which of these do you think is clearer for the reader: (extract from test code)

```hs
Selection.verify constraints selection === SelectionCorrect
```

```hs
Selection.verify constraints selection === Right ()
```

I think the first example is much clearer, as it doesn't require the reader to understand what `Right ()` means in this particular context.

> Also, looking now, I wonder if having the correctness checks in this module actually makes sense. Maybe could be moved later, if an opportunity arises.

The `Selection.create`, `Selection.extend` and `Selection.balance` functions all refer to `Selection.verify` in their documentation:

```hs
-- | Creates a selection with the given inputs.
--
-- Guarantees the following property for a returned selection 's':
--
-- >>> verify s == SelectionCorrect
```
This gives a simple, concise statement of what these functions guarantee, without having to list all the properties that the resultant selection should satisfy. The reader can navigate to `verify`, and see for themselves.

If we were to move the `verify` function into the test suite, we'd presumably have two options:

1. Within the comments for `create`, `extend` and `balance`, list all the properties that they satisfy. But there are many such properties. And we'd have to write them out three times, or refer to a shared comment. Comments are not compiler checked, and it's likely that they'd bit rot over time.

2. Within the comments for `create`, `extend` and `balance`, tell the reader to look at the properties in `SelectionSpec`. But this arguably increases the burden of understanding for the reader, as they now have to navigate to a different module in a different subproject just to understand the properties that these functions guarantee.

jonathanknowles replied:
@Anviking We actually already use a similar technique within `UTxOIndex`:

https://github.com/input-output-hk/cardano-wallet/blob/2acc0305e51d5b725701897b96cbc1ed8a61a477/lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs#L490-L501

https://github.com/input-output-hk/cardano-wallet/blob/2acc0305e51d5b725701897b96cbc1ed8a61a477/lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs#L504-L506

jonathanknowles committed:
Use `forAll` to simplify `PlanningSpec`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r616047229

jonathanknowles committed:
Use `forAll` to simplify `SelectionSpec`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r616047229

jonathanknowles committed:
Use `forAll` to simplify `MigrationSpec`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r616047229

jonathanknowles committed:
Provide a shared definition for `genRewardWithdrawal`.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -0,0 +1,956 @@
+--
+-- Use 'create' to create a selection with one or more inputs.
+-- Use 'extend' to extend a selection with an additional input.
+-- Use 'verify' to verify the correctness of a selection.
```
```suggestion
```

This makes it sound like you should be using `verify` when using this module, when in reality it is used only for testing. 

I think if you simply drop this line, the module doc becomes really neat

jonathanknowles replied:
I agree that removing this line wouldn't be harmful.

In any case, if the user navigates to `create` or `extend`, they'll soon find out that the comments for these functions refer to `verify` as a way to describe what they guarantee.

jonathanknowles replied:
Fixed in 7bd44365061e0dfc9d435a0ffcb34c12d1dc8fa5.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -0,0 +1,956 @@
+     (SelectionSizeIncorrectError size)
+    deriving (Eq, Show)
+
+-- | Verifies a selection for correctness.
```
How about just adding:
```suggestion
-- | Verifies a selection for correctness. For testing, as selections from @create@ and @extend@ are guaranteed to be correct.
```
to make it clear that you're not normally supposed to care about it.

Also I forgot exactly the rationale for having it in the library code, but _maybe_ you could add a quick word about that.

jonathanknowles replied:
> to make it clear that you're not normally supposed to care about it.

Thanks for this suggestion. I've added a comment in ec6970e420525f13d223a70b33eb3f81a3fb6f46.

> Also I forgot exactly the rationale for having it in the library code, but maybe you could add a quick word about that.

See: https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r618171707

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -493,3 +503,66 @@ txOutMinTokenQuantity = TokenQuantity 1
+-- This allows parts of a transaction to be costed individually, without
+-- having to compute the cost of an entire transaction.
+--
+data TxConstraints s = TxConstraints
```
This type will never be used when say listing Txs from DB.

It's currently used from Migration.Selection, Migration.Planning and Migration.

Is not Migration.Selection a better module for it?

Anviking replied:
Although having it next to the similar `TokenBundleSizeAssessor` makes sense, so maybe keep it.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -493,3 +503,66 @@ txOutMinTokenQuantity = TokenQuantity 1
+data TxConstraints s = TxConstraints
+    { txBaseCost :: Coin
+      -- ^ The constant cost of an empty transaction.
+    , txBaseSize :: s
```
Suggestion: to get rid of some boilerplate, you _could_ do:

```haskell
data SizeAndCost s = SizeAndCost
    { size :: s
    , cost :: Coin
    }

data TxConstraints s = TxConstraints
    { txBase :: SizeAndCost s
      -- ^ The constant size and cost of an empty transaction.
    , txInput :: SizeAndCost s
      -- ^ The size and cost of including an extra input to a transaction.
    , txOutput :: TokenBundle -> SizeAndCost s
      -- ^ The size and cost of including an extra output to a transaction.
    , txRewardWithdrawal :: Coin -> SizeAndCost s
      -- ^ The variable size and cost of a reward withdrawal.
    , txOutputMaximumSize :: s
      -- ^ The maximum size allowed
    , txOutputMaximumTokenQuantity :: TokenQuantity
      -- ^ The maximum token quantity that can appear in a transaction output.
    , txOutputMinimumAdaQuantity :: TokenMap -> Coin
      -- ^ The variable minimum ada quantity of a transaction output.
    , txMaximumSize :: s
      -- ^ The maximum size of a transaction.
    }
```

jonathanknowles replied:
I agree this would reduce the number of fields, but I'm not entirely convinced the benefit would exceed the cost!

In particular, I find the revised field names to be less clear. For example, at first glance, `txInput` seems like it ought to return an input, but actually it's the cost and size **_of_** an input. IMO, `txInputSize` and `txInputCost` seem more straightforward and easy to understand.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Planning.hs:
```diff
@@ -0,0 +1,307 @@
+-- for themselves. A "supporter" entry is only added to the selection if there
+-- is not enough ada to pay for a "freerider" entry.
+--
+extendSelection
```
```suggestion
fillSelection
```
?
Otherwise I would have thought it only performed one step.

jonathanknowles replied:
Thanks for the suggestion to rename this. I agree it could be a little clearer.

I've renamed this to `extendSelectionUntilFull`, as it does indeed repeatedly extend a selection until it's full.

Fixed in 97f5c261b6d23b65e14dcb67c72101a62db8b58d.

Anviking reviewed (commented):
Some stylistic comments; will review more

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -493,3 +503,66 @@ txOutMinTokenQuantity = TokenQuantity 1
+    , txBaseSize :: s
+      -- ^ The constant size of an empty transaction.
+    , txInputCost :: Coin
+      -- ^ The constant cost of a transation input.
```
```suggestion
      -- ^ The constant cost of a transation input.
      -- 
      -- NOTE: For UTxOs sharing the same payment key this function and the current wallet in general will slightly overestimate the cost / fee, as it assumes one witness per input. (https://github.com/input-output-hk/cardano-wallet/blob/2acc0305e51d5b725701897b96cbc1ed8a61a477/lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs#L582-L584)
```

I think this deserves mentioning

jonathanknowles replied:
Good point!

I've added a comment in a95d51bb3726359c971ae62b92379522ff38c05e.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -0,0 +1,956 @@
+        { outputBundle :: TokenBundle
+          -- ^ The output that is below the expected minimum ada quantity.
+        , expectedMinimumAdaQuantity :: Coin
+          -- ^ The expected minimum ada quantity.
```
Tbh, wondering if all of this error-boilerplate is worth it. Couldn't `verify` simply be an `Either String ()` in the spec, or something like that?

(Similar to https://github.com/input-output-hk/cardano-wallet/pull/2618/files#r617676467)

jonathanknowles committed:
Improve the output of test failures.

This commit adds two utility functions, and uses them to improve the
output of test failures:

  * report:

    Adds a named variable to the counterexample output of a property.
    In the event of a test failure, a pretty-printed representation of
    the variable is shown as part of the counterexample output.

  * verify:

    Adds a named test condition to a property. In the event of a test
    failure, the names of all the conditions that failed are added to
    the counterexample output.

Sample output:

```
    test/unit/Cardano/Wallet/Primitive/Migration/SelectionSpec.hs:146:9:
    1) Minimizing fees, prop_minimizeFeeStep
         Falsified (after 1 test):
           (*)

           "mockConstraints":
               MockTxConstraints
                   { mockTxCostFunction = MockTxCostFunction
                   { baseCost = Coin 643
                   , sizeCost = Coin 2
                   }
                   , mockTxBaseSize = 835
                   , mockTxInputSize = 3
                   , mockTxOutputMaximumSize = 400
                   , mockTxOutputMaximumTokenQuantity = 1077
                   , mockTxOutputMinimumAdaQuantity =
                   MockTxOutputMinimumAdaQuantity
                      { perOutput = Coin 6
                      , perOutputAsset = Coin 1
                      }
                   , mockTxMaximumSize = 2447
                   }

           "feeExcessBefore":
               Coin 5335

           "feeExcessAfter":
               Coin 5536

           "feeExcessDifference":
               Coin 1

           Condition violated: "feeExcessAfter <= feeExcessBefore"
```

jonathanknowles committed:
Remove link to `verify` function from `Selection` module documentation.

The `verify` function is part of the public interface and documentation,
but ordinary users of this module are not generally expected to call it
directly (unless they're writing tests).

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r617449171

jonathanknowles committed:
Emphasise that `Selection.verify` is provided as a convenience for testing.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r617454660

jonathanknowles committed:
Rename `extendSelection` to `extendSelectionUntilFull`.

This makes the intended behaviour (of repeatedly extending a selection
until its full) a bit clearer.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r617677378

jonathanknowles committed:
Clarify the assumptions of `txInputCost` and `txInputSize`.

These functions both assume that one witness is required per input.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#discussion_r617699319

jonathanknowles committed:
Make the order consistent for module `Migration.Selection`.

This commit makes the order consistent between:

- The order in which functions are listed in the export list
- The order in which functions are actually defined.

jonathanknowles committed:
Check that `Selection.create` and `Selection.extend` minimize the fee.

It's really important that both of these functions minimize the fee.

So we check that:

1. Calling `balance` on a resultant selection leaves it unchanged.
2. Calling `minimizeFee` on a resultant fee excess leaves it unchanged.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2618#issuecomment-822413404

jonathanknowles committed:
Strengthen the `totalFee` check within `Planning.prop_createPlan`.

Before this commit, to check the correctness of `totalFee`, we simply
summed the `fee` field values of the individual selections, and checked
this value matched the total.

Now, we manually compute the expected total fee by finding the
difference between:

    1. the total amount of ada provided by the inputs and reward balance;
    2. the total amount of ada spent by the outputs.

This is a stronger test, as we avoid relying on the `Selection` module's
guarantee that the fee of each selection is correct.

Anviking reviewed (approved):
LGTM.

jonathanknowles committed:
Check that `createPlan` preserves the total non-ada token balance.
